### PR TITLE
[TEST] Add test for retrying replica operations with real network

### PR DESCRIPTION
**Pull request by dakrone**
Related to #24745

@jasontedor as we discussed, this has the following three properties:

1. Passes before inlining checkpoint change, I backported the test to 5.x to
   make sure it passes.

2. Fails after inlining checkpoint change, I undid your fix for the Shard ->
   Replica request and it does indeed fail.

3. Passes after your one-line change ConcreteShardRequest -> ConcreteReplicaRequest,
   this does now pass, now that the correct class is used for the request (fixed
   as part of 657686cefb736999b2231b4caee318da14e9462a)

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 1]
```diff
@@ -918,6 +930,60 @@ protected ReplicaResult shardOperationOnReplica(Request request, IndexShard repl
...
+    public void testRetryOnReplicaWithRealTransport() throws Exception {
+        final ShardId shardId = new ShardId("test", "_na_", 0);
+        ClusterState state = state(shardId.getIndexName(), true, ShardRoutingState.STARTED, ShardRoutingState.STARTED);
+        final ShardRouting replica = state.getRoutingTable().shardRoutingTable(shardId).replicaShards().get(0);
+        // simulate execution of the node holding the replica
+        state = ClusterState.builder(state).nodes(DiscoveryNodes.builder(state.nodes()).localNodeId(replica.currentNodeId())).build();
+        setState(clusterService, state);
+        AtomicBoolean throwException = new AtomicBoolean(true);
+        final ReplicationTask task = maybeTask();
+        Map<String, TransportAddress> poorMansDNS = new ConcurrentHashMap<>();
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: I don't see this used anywhere?

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 2]
```diff
@@ -918,6 +930,60 @@ protected ReplicaResult shardOperationOnReplica(Request request, IndexShard repl
         assertConcreteShardRequest(capturedRequest.request, request, replica.allocationId());
     }
 
+    public void testRetryOnReplicaWithRealTransport() throws Exception {
+        final ShardId shardId = new ShardId("test", "_na_", 0);
+        ClusterState state = state(shardId.getIndexName(), true, ShardRoutingState.STARTED, ShardRoutingState.STARTED);
+        final ShardRouting replica = state.getRoutingTable().shardRoutingTable(shardId).replicaShards().get(0);
+        // simulate execution of the node holding the replica
+        state = ClusterState.builder(state).nodes(DiscoveryNodes.builder(state.nodes()).localNodeId(replica.currentNodeId())).build();
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: Can we get away with not reusing this variable, instead having two different variables with clearer names?

**Review by jasontedor — CHANGES_REQUESTED** [review]
Thank you @dakrone, this looks good, I left a couple of minor comments.

**Comment by dakrone** [comment]
Thanks for taking a look @jasontedor! I pushed a commit addressing your comments

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**